src/qubx/core/strategy.py

```diff
@@ -133,719 +124,119 @@ def is_simulated_trading(self) -> bool:
         """
-        if not self._cache.is_data_ready():
-            return None
+        raise NotImplementedError("get_position_size is not implemented")
 
-        # times are in seconds here
-        prev_fit_time, now_fit_time = data
 
-        # - we need to run this in separate thread
-        self.__fit_is_running = True
-        self._run_in_thread_pool(
-            self._invoke_on_fit,
-            (pd.Timestamp(now_fit_time, unit="s"), pd.Timestamp(prev_fit_time, unit="s") if prev_fit_time else None),
-        )
+class PositionsTracker:
+    """
+    Tracks position and processing signals. It can contains logic for risk management for example.
+    """
 
-        return None
+    _sizer: IPositionSizer
 
-    def _processing_hist_bar(self, symbol: str, bar: Bar) -> TriggerEvent | None:
-        # - processing single historical bar
-        #   here it just updates cache - historical bar can't trigger strategy logic
-        self._cache.update_by_bar(symbol, bar)
-        return None
+    def __init__(self, sizer: IPositionSizer) -> None:
+        self._sizer = sizer
 
-    def _processing_hist_bars(self, symbol: str, bars: List[Bar]) -> TriggerEvent | None:
-        # - processing historical bars as list
-        for b in bars:
-            self._processing_hist_bar(symbol, b)
-        return None
+    def get_position_sizer(self, instrument: Instrument) -> IPositionSizer:
+        return self._sizer
 
-    def _check_if_need_trigger_on_bar(self, symbol: str, bar: Bar | None):
-        if self._trig_on_bar:
-            t = self.trading_service.time().item()
-            _time_to_trigger = t % self._trig_bar_freq_nsec >= self._trig_interval_in_bar_nsec
-            if _time_to_trigger:
-                # we want to trigger only first one - not every
-                if not self._current_bar_trigger_processed:
-                    self._current_bar_trigger_processed = True
-                    return TriggerEvent(self.time(), "bar", self._symb_to_instr.get(symbol), bar)
-            else:
-                self._current_bar_trigger_processed = False
-        return None
-
-    @_SW.watch("StrategyContext")
-    def _processing_bar(self, symbol: str, bar: Bar) -> TriggerEvent | None:
-        # - processing current bar's update
-        self._cache.update_by_bar(symbol, bar)
+    def process_signals(self, ctx: StrategyContext, gathering: IPositionGathering, signals: List[Signal]):
+        """
+        By default it just treat signals as pure positions sizes
+        """
+        for s in signals:
+            new_position = self.get_position_sizer(s.instrument).get_position_size(ctx, s)
+            gathering.alter_position_size(ctx, s.instrument, new_position, s.price)
 
-        # - check if it's time to trigger the on_event if it's configured
-        return self._check_if_need_trigger_on_bar(symbol, bar)
+    def update(self, ctx: StrategyContext, instrument: Instrument, update: Quote | Trade | Bar): ...
```

Comment:
I don't see yet who is responsible for calling this? Will it be wired up in the strategy context? Or should the strategy call this itself from the `on_event`?

Reply:
This called from StrategyContex when new market data received. see _process_bar/quote/trade methods